File: backend/services/shipment_service.go
package services
import (
	"errors"
	"fmt"
	"time"
	"cronos-backend/blockchain"
	"cronos-backend/models"
	"cronos-backend/repositories"
	"github.com/google/uuid"
)
type ShipmentService struct {
	ShipmentRepo *repositories.ShipmentRepository
	OrderRepo    *repositories.OrderRepository
	Ledger       *blockchain.Ledger
}
func NewShipmentService(shipmentRepo *repositories.ShipmentRepository, orderRepo *repositories.OrderRepository, ledger *blockchain.Ledger) *ShipmentService {
	return &ShipmentService{ShipmentRepo: shipmentRepo, OrderRepo: orderRepo, Ledger: ledger}
}
type CreateShipmentRequest struct {
	OrderID           uuid.UUID  `json:"order_id" validate:"required"`
	CourierID         uuid.UUID  `json:"courier_id" validate:"required"`
	TrackingNumber    string     `json:"tracking_number"`
	EstimatedDelivery *time.Time `json:"estimated_delivery"`
}
func (s *ShipmentService) CreateShipment(req CreateShipmentRequest) (*models.Shipment, error) {
	order, err := s.OrderRepo.FindByID(req.OrderID)
	if err != nil {
		return nil, errors.New("order not found")
	}
	if order.Status != "paid" && order.Status != "processing" && order.Status != "pending" {
		return nil, errors.New("order must be paid or pending before shipping")
	}
	shipment := &models.Shipment{
		OrderID:           req.OrderID,
		CourierID:         &req.CourierID,
		TrackingNumber:    req.TrackingNumber,
		Status:            "pending",
		EstimatedDelivery: req.EstimatedDelivery,
	}
	if err := s.ShipmentRepo.Create(shipment); err != nil {
		return nil, errors.New("failed to create shipment")
	}
	order.Status = "processing"
	s.OrderRepo.Update(order)
	return s.ShipmentRepo.FindByID(shipment.ID)
}
func (s *ShipmentService) UpdateShipmentStatus(courierID, shipmentID uuid.UUID, status, location, notes string) (*models.Shipment, error) {
	shipment, err := s.ShipmentRepo.FindByID(shipmentID)
	if err != nil {
		return nil, errors.New("shipment not found")
	}
	if shipment.CourierID == nil || *shipment.CourierID != courierID {
		return nil, errors.New("unauthorized")
	}
	validTransitions := map[string][]string{
		"pending": {"pickup"},
		"pickup":  {"transit"},
		"transit": {"delivered"},
	}
	allowed, ok := validTransitions[shipment.Status]
	if !ok {
		return nil, errors.New("shipment already in final state")
	}
	valid := false
	for _, s := range allowed {
		if s == status {
			valid = true
			break
		}
	}
	if !valid {
		return nil, fmt.Errorf("invalid status transition from %s to %s", shipment.Status, status)
	}
	shipment.Status = status
	switch status {
	case "delivered":
		now := time.Now()
		shipment.ActualDelivery = &now
		order, _ := s.OrderRepo.FindByID(shipment.OrderID)
		order.Status = "delivered"
		s.OrderRepo.Update(order)
	case "pickup":
		order, _ := s.OrderRepo.FindByID(shipment.OrderID)
		order.Status = "shipped"
		s.OrderRepo.Update(order)
	}
	s.ShipmentRepo.Update(shipment)
	log := &models.ShipmentLog{
		ShipmentID: shipmentID,
		Status:     status,
		Location:   location,
		Notes:      notes,
		Timestamp:  time.Now(),
	}
	s.ShipmentRepo.CreateLog(log)
	eventType := "shipment_" + status
	s.Ledger.RecordEvent(eventType, courierID, "shipment", shipment.OrderID, map[string]interface{}{
		"status":   status,
		"location": location,
	})
	return s.ShipmentRepo.FindByID(shipmentID)
}
func (s *ShipmentService) GetMyShipments(courierID uuid.UUID) ([]models.Shipment, error) {
	return s.ShipmentRepo.FindByCourierID(courierID)
}
func (s *ShipmentService) GetShipment(id uuid.UUID) (*models.Shipment, error) {
	return s.ShipmentRepo.FindByID(id)
}
func (s *ShipmentService) GetShipmentLogs(shipmentID uuid.UUID) ([]models.ShipmentLog, error) {
	return s.ShipmentRepo.FindLogsByShipmentID(shipmentID)
}
func (s *ShipmentService) GetAllShipments() ([]models.Shipment, error) {
	return s.ShipmentRepo.FindAll()
}

File: backend-spring/src/main/java/com/cronos/service/ShipmentService.java
package com.cronos.service;

import com.cronos.blockchain.BlockchainLedger;
import com.cronos.dto.request.CreateShipmentRequest;
import com.cronos.entity.*;
import com.cronos.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDateTime;
import java.util.*;

@Service
@RequiredArgsConstructor
public class ShipmentService {
    private final ShipmentRepository shipmentRepo;
    private final ShipmentLogRepository shipmentLogRepo;
    private final OrderRepository orderRepo;
    private final BlockchainLedger ledger;

    public Shipment createShipment(CreateShipmentRequest req) {
        CronosOrder order = orderRepo.findById(req.getOrderId()).orElseThrow(() -> new RuntimeException("order not found"));
        if (!"paid".equals(order.getStatus()) && !"processing".equals(order.getStatus()) && !"pending".equals(order.getStatus()))
            throw new RuntimeException("order must be paid or pending before shipping");

        Shipment shipment = Shipment.builder().orderId(req.getOrderId()).courierId(req.getCourierId())
                .trackingNumber(req.getTrackingNumber()).status("pending").estimatedDelivery(req.getEstimatedDelivery()).build();
        shipmentRepo.save(shipment);
        order.setStatus("processing"); orderRepo.save(order);
        return shipmentRepo.findById(shipment.getId()).get();
    }

    @Transactional
    public Shipment updateShipmentStatus(UUID courierId, UUID shipmentId, String status, String location, String notes) {
        Shipment shipment = shipmentRepo.findById(shipmentId).orElseThrow(() -> new RuntimeException("shipment not found"));
        if (shipment.getCourierId() == null || !shipment.getCourierId().equals(courierId)) throw new RuntimeException("unauthorized");

        Map<String, List<String>> validTransitions = Map.of("pending", List.of("pickup"), "pickup", List.of("transit"), "transit", List.of("delivered"));
        List<String> allowed = validTransitions.get(shipment.getStatus());
        if (allowed == null) throw new RuntimeException("shipment already in final state");
        if (!allowed.contains(status)) throw new RuntimeException("invalid status transition from " + shipment.getStatus() + " to " + status);

        shipment.setStatus(status);
        if ("delivered".equals(status)) {
            shipment.setActualDelivery(LocalDateTime.now());
            CronosOrder order = orderRepo.findById(shipment.getOrderId()).get();
            order.setStatus("delivered"); orderRepo.save(order);
        } else if ("pickup".equals(status)) {
            CronosOrder order = orderRepo.findById(shipment.getOrderId()).get();
            order.setStatus("shipped"); orderRepo.save(order);
        }
        shipmentRepo.save(shipment);

        ShipmentLog log = ShipmentLog.builder().shipmentId(shipmentId).status(status).location(location).notes(notes).timestamp(LocalDateTime.now()).build();
        shipmentLogRepo.save(log);

        ledger.recordEvent("shipment_" + status, courierId, "shipment", shipment.getOrderId(), Map.of("status", status, "location", location != null ? location : ""));
        return shipmentRepo.findById(shipmentId).get();
    }

    public List<Shipment> getMyShipments(UUID courierId) { return shipmentRepo.findByCourierIdOrderByCreatedAtDesc(courierId); }
    public List<Shipment> getAllShipments() { return shipmentRepo.findAllByOrderByCreatedAtDesc(); }
    public List<ShipmentLog> getShipmentLogs(UUID shipmentId) { return shipmentLogRepo.findByShipmentIdOrderByTimestampDesc(shipmentId); }
}
